experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl

```diff
@@ -0,0 +1,326 @@
+  return length(C) == 1 ? (1:states(M)[C[1]]) : Iterators.product([1:states(M)[i] for i in C]...)
+end
+
+@doc raw"""
+  @attr Tuple{MPolyRing, GenDict} parameter_ring(M::DiscreteGraphicalModel{Graph{Undirected}, T}; cached=false)
+
+Return the ring of parameters of the statistical model, together with a
+Dict for indexing its generators.
+
+``` jldoctest 
+julia> M = discrete_graphical_model(graph_from_edges([[1,2], [2,3]]), [2,2,2])
+Discrete Graphical Model on a Undirected graph with 3 nodes and 2 edges with states [2, 2, 2]
+
+julia> _, PR_gens = parameter_ring(M);
+
+julia> PR_gens[[1, 2], (2, 2)]
+t[1, 2](2, 2)
+```
+"""
+@attr Tuple{
+  MPolyRing,
+  GenDict
+} function parameter_ring(M::DiscreteGraphicalModel{Graph{Undirected}, T}; cached=false) where T
+  cliques = sort(sort.(collect.(maximal_cliques(graph(M)))))
+  Xs = [sort(vec(collect(state_space(M, C)))) for C in cliques]
+  params = [(C, x) for (C, X) in Iterators.zip(cliques, Xs) for x in X]
+  gen_names = sort([varnames(M)[:t] * string(C) * string(x) for (C, x) in params])
+  R, t = polynomial_ring(QQ, gen_names; cached=cached)
+  gens_dict = Dict((C, x) => t[i] for (i, (C, x)) in enumerate(params))
+  return (R, gens_dict)
+end
+
+@doc raw"""
+    vanishing_ideal(M::DiscreteGraphicalModel{Graph{Undirected}, L} where L
+
+The vanishing ideal is toric of an undirected discrete graphical model
+is toric. Therefore its `vanishing_ideal` method defaults to `algorithm=:markov`.
+"""
+function vanishing_ideal(M::DiscreteGraphicalModel{Graph{Undirected}, L}) where L
+  vanishing_ideal(M; algorithm=:markov)
+end
+
+@doc raw"""
+  parameterization(M::DiscreteGraphicalModel{Graph{Undirected}, L})
+
+Creates the polynomial map which parameterizes the vanishing ideal of the
+undirected discrete graphical model `M`. It sends each probability generator
+`p[s]` of the model ring to the product over all generators `t[C](sp)` of
+the parameter ring where `s[C] == sp`, where `C` is a maximal clique of the
+graph and `sp` is a marginal state of the random vector.
+
+## Examples
+
+``` jldoctest
+julia> M = discrete_graphical_model(graph_from_edges([[1,2], [2,3]]), [2,2,2])
+Discrete Graphical Model on a Undirected graph with 3 nodes and 2 edges with states [2, 2, 2]
+
+julia> parametrization(M)
+Ring homomorphism
+  from multivariate polynomial ring in 8 variables over QQ
+  to multivariate polynomial ring in 8 variables over QQ
+defined by
+  p[1,1,1] -> t[1, 2](1, 1)*t[2, 3](1, 1)
+  p[2,1,1] -> t[1, 2](2, 1)*t[2, 3](1, 1)
+  p[1,2,1] -> t[1, 2](1, 2)*t[2, 3](2, 1)
+  p[2,2,1] -> t[1, 2](2, 2)*t[2, 3](2, 1)
+  p[1,1,2] -> t[1, 2](1, 1)*t[2, 3](1, 2)
+  p[2,1,2] -> t[1, 2](2, 1)*t[2, 3](1, 2)
+  p[1,2,2] -> t[1, 2](1, 2)*t[2, 3](2, 2)
+  p[2,2,2] -> t[1, 2](2, 2)*t[2, 3](2, 2)
+```
+"""
+function parametrization(M::DiscreteGraphicalModel{Graph{Undirected}, L}) where L
+  S, pd = model_ring(M)
+  R, td = parameter_ring(M)
+  images = []
```

Comment:
type unstable, between would be
```suggestion
  images = SomeType[]
```
where presumably `SomeType` is the type of the values stored in `td` -- hopefully this can be derived from the type of `M` plus `mpoly_type(QQFieldElem)`  ?

Reply:
`elem_type(R)[]` should be fine here no?